platforms/chibios/drivers/ws2812_pwm.c

```diff
@@ -29,6 +29,16 @@
 #    error "please consult your MCU's datasheet and specify in your config.h: #define WS2812_DMAMUX_ID STM32_DMAMUX1_TIM?_UP"
 #endif
 
+/* Summarize https://www.st.com/resource/en/application_note/an4013-stm32-crossseries-timer-overview-stmicroelectronics.pdf to
+ * figure out if we are using a 32bit timer. This is needed to setup the DMA controller correctly.
+ * Ignore STM32H7XX and STM32U5XX as they are not supported by ChibiOS.
+ */
+#if (WS2812_PWM_DRIVER == PWMD2) && (!defined(STM32F1XX) || !defined(STM32L0XX) || !defined(STM32L1XX))
+#    define WS2812_PWM_TIMER_32BIT
+#elif (WS2812_PWM_DRIVER == PWMD5) && (!defined(STM32F1XX))
+#    define WS2812_PWM_TIMER_32BIT
+#endif
```

Comment:
The `WS2812_PWM_DRIVER == PWMD2` comparison does not work at the preprocessor level (looks like it is always evaluated as true, so the resulting code still works, but the optimization for 16-bit timers on F4xx never gets enabled). Also the logic in the second part is wrong (it needs `&&`, not `||`).

The following code seems to work for me:

```suggestion
#if !defined(STM32F1XX) && !defined(STM32L0XX) && !defined(STM32L1XX)
#    define WS2812_PWM_TIMER_32BIT_PWMD2 1
#endif
#if !defined(STM32F1XX)
#    define WS2812_PWM_TIMER_32BIT_PWMD5 1
#endif
#define WS2812_CONCAT1(a, b) a##b
#define WS2812_CONCAT(a, b) WS2812_CONCAT1(a, b)
#if WS2812_CONCAT(WS2812_PWM_TIMER_32BIT_, WS2812_PWM_DRIVER)
#    define WS2812_PWM_TIMER_32BIT
#endif
```

Reply:
After a good conversation I learned that comparison in defines in the C preprocessor will only work correctly for numerical values. Proposed code seems to be the cleanest.